### PR TITLE
Add ExecutionModeNamedBarrierCountINTEL

### PR DESCRIPTION
**Pull request by vmustya**

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -4560,6 +4560,12 @@ bool LLVMToSPIRVBase::transExecutionMode() {
           break;
         AddSingleArgExecutionMode(static_cast<ExecutionMode>(EMode));
       } break;
+      case spv::internal::ExecutionModeNamedBarrierCountINTEL: {
+        unsigned NBarrierCnt;
+        N.get(NBarrierCnt);
+        BF->addExecutionMode(new SPIRVExecutionMode(
+            BF, static_cast<ExecutionMode>(EMode), NBarrierCnt));
+      } break;
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:4568]: Here should be a check, if the appropriate extension is enabled.
Also we need to add the appropriate capability.

**Comment by MrSidims** [comment]
@vmustya hi, where can we ready about this extension? Also the test for the feature is missing.

**Comment by vmustya** [comment]
> @vmustya hi, where can we ready about this extension? Also the test for the feature is missing.

This stuff is a part of SPV_INTEL_vector_compute extension: https://github.com/intel/llvm/blob/68def46184daa35d7819793b3a3eed733ee27137/sycl/doc/extensions/SPIRV/SPV_INTEL_vector_compute.asciidoc

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -4560,6 +4560,16 @@ bool LLVMToSPIRVBase::transExecutionMode() {
           break;
         AddSingleArgExecutionMode(static_cast<ExecutionMode>(EMode));
       } break;
+      case spv::internal::ExecutionModeNamedBarrierCountINTEL: {
+        if (!BM->isAllowedToUseExtension(ExtensionID::SPV_INTEL_vector_compute))
+          break;
+        unsigned NBarrierCnt;
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:4566]: nitpick:
```suggestion
        unsigned NBarrierCnt = 0;
```

File: test/transcoding/SPV_INTEL_vector_compute/exec_mode_named_barriers_count.ll [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+; RUN: FileCheck %s --input-file %t.ll  -check-prefix=LLVM
+
+; ModuleID = 'slm.bc'
+source_filename = "slm.cpp"
+target datalayout = "e-p:64:64-i64:64-n8:16:32"
+target triple = "spir"
+
+; LLVM-DAG: @k_rte{{[^a-zA-Z0-9_][^#]*}}#[[K_RTE:[0-9]+]]
+; LLVM-DAG: attributes #[[K_RTE]]{{[^0-9].*"VCNamedBarrierCount"="8"}}
+; SPV: EntryPoint {{[0-9]+}} [[S_RTE:[0-9]+]] "k_rte"
```
Comment (MrSidims) [inline test/transcoding/SPV_INTEL_vector_compute/exec_mode_named_barriers_count.ll:16]: Please also check if the appropriate OpExtension and OpCapability were added

**Review by MrSidims — COMMENTED** [review]
LGTM with 2 nits

File: lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (vmustya) [reply lib/SPIRV/SPIRVWriter.cpp]: Done

File: test/transcoding/SPV_INTEL_vector_compute/exec_mode_named_barriers_count.ll (hunk 3 above)
Comment (vmustya) [reply test/transcoding/SPV_INTEL_vector_compute/exec_mode_named_barriers_count.ll:18]: Done

File: lib/SPIRV/PreprocessMetadata.cpp [hunk 4]
```diff
@@ -348,6 +348,20 @@ void PreprocessMetadataBase::preprocessVectorComputeMetadata(Module *M,
           .add(spv::internal::ExecutionModeFastCompositeKernelINTEL)
           .done();
     }
+
+    if (Attrs.hasFnAttribute(kVCMetadata::VCNamedBarrierCount)) {
+      SPIRVWord NBarrierCnt = 0;
+      Attrs
+          .getAttribute(AttributeList::FunctionIndex,
+                        kVCMetadata::VCNamedBarrierCount)
```
Comment (MrSidims) [inline lib/SPIRV/PreprocessMetadata.cpp:356]:
```suggestion
          .getFnAttr(kVCMetadata::VCNamedBarrierCount)
```

File: lib/SPIRV/PreprocessMetadata.cpp [hunk 5]
```diff
@@ -348,6 +348,20 @@ void PreprocessMetadataBase::preprocessVectorComputeMetadata(Module *M,
           .add(spv::internal::ExecutionModeFastCompositeKernelINTEL)
           .done();
     }
+
+    if (Attrs.hasFnAttribute(kVCMetadata::VCNamedBarrierCount)) {
```
Comment (MrSidims) [inline lib/SPIRV/PreprocessMetadata.cpp:352]:
```suggestion
    if (Attrs.hasFnAttr(kVCMetadata::VCNamedBarrierCount)) {
```

**Review by MrSidims — APPROVED** [review]

**Pull request merged**